feat(databricks): add basic op implementations

rexledesma opened the pull request:
### Summary & Motivation
There are two standard ways of triggering jobs on Databricks: either you can submit API calls to trigger an existing job, or you can construct a job definition on the fly. Here, we add support for both methods.

### How I Tested These Changes
pytest, local

rexledesma commented:
Current dependencies on/for this PR:
* master
  * **PR #12364** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12364" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #12366** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12366" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #12377** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12377" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #12382** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12382" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
          * **PR #12492** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12492" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12492?utm_source=stack-comment).

sryza reviewed (commented):
What do you think about making these op factories to make them consistent with `create_databricks_job_op`?

Aside from consistency, it also makes it easier to customize the name and number of inputs.

rexledesma commented:
> What do you think about making these op factories to make them consistent with `create_databricks_job_op`?

I actually do not like this pattern: I believe asset/op factories should be an abstraction that the user defines, since the amount of customization that they want here could be unbounded. 

> Aside from consistency, it also makes it easier to customize the name and number of inputs.

- For the name, could we direct the user to use `.alias`? 
- For the number of inputs, can't `start_after` take in a list of inputs? Since this is a `Nothing` input anyways, one input will suffice in order to define dependency order from `N` inputs.

rexledesma commented:
@sryza After some more thought, I'll wrap these ops in a factory pattern. But I'll keep the arguments of the factory to be arbitrary kwargs that are passed down to the op. This accomplishes two things. This way, the factory pattern is infinitely customizable out of the box, but retains the sensible defaults of the underlying op.

Out-of-the-box factories like this are a little complicated since there's another layer of indirection that the user needs to understand, but this plays nicely for interpreting arbitrary assets from the job using `AssetsDefintion.from_op`.

```
def factory(**kwargs):
    @op(**{**sensible_defaults, **kwargs})
    def _impl():
        pass
    
    return _impl
```

rexledesma commented:
Changed a couple of things here:

- refactored to op factory notation
- lifted up required configuration to be parameters of the op factory, making `.configured` optional

sryza commented on python_modules/libraries/dagster-databricks/dagster_databricks/ops.py:
```diff
@@ -135,3 +136,239 @@ def create_ui_url(databricks_client, op_config):
+    override parameters for the Databricks Job.
+
+    Arguments:
+        databricks_job_id (int): The ID of the Databricks Job to be executed.
```
I know you mentioned it above, but I think it's worth mentioning the kwargs here and being specific about what kinds of parameters are accepted (all the parameters to the op decorator?)

sryza commented on python_modules/libraries/dagster-databricks/dagster_databricks/ops.py:
```diff
@@ -135,3 +136,239 @@ def create_ui_url(databricks_client, op_config):
         )
+
+
+def create_databricks_run_now_op(databricks_job_id: int, **kwargs) -> OpDefinition:
```
I wonder if this should take a `databricks_job_configuration` like the other function? Seems nice to not need to use `configured` to specify `python_params`?

sryza reviewed (commented):
This looks close to me - just left one comment and one question

rexledesma commented:
@sryza Great - that feedback should be addressed in https://github.com/dagster-io/dagster/pull/12492/commits/aa7ba1752efc4c65ba678c548a9092e619c46b01. The user shouldn't have to use `configured` at all now, and will still get the benefits of runtime configuration via the launchpad.

schrockn commented:
My general impression is that this API offers too many "degrees of freedom" here by passing through all the keyword args.

sryza reviewed (approved):
Nice nice nice

sryza commented on python_modules/libraries/dagster-databricks/dagster_databricks/ops.py:
```diff
@@ -135,3 +138,247 @@ def create_ui_url(databricks_client, op_config):
         )
+
+
+def create_databricks_run_now_op(
```
Relevant discussion here btw: https://elementl-workspace.slack.com/archives/C04JNJCSS13/p1677526031428709

sryza replied:
Maybe worth a chat? I agree that this is an API that's hard to walk back and I'm a little nervous about it, but I also think it's a promising pattern.  

Maybe best to be conservative and leave out the runtime config for now?

Having two separate top-level APIs seems costly to me - I think many users will find the distinction immaterial and be deterred by needing to make a choice.

rexledesma replied:
I think the confusion comes from the underlying Databricks API call.

When using the `Run Now` API, you can specify a custom list of params to pass into the underlying Databricks Job. Otherwise, the default params of the Databricks Job will be used.

Similarly, when using the `Submit Run` Databricks API, you can specify the list of Databricks tasks you want to run, as well as the configuration of the cluster on which the tasks run. It is this cluster configuration that could be useful to configure at run time, but from talking to some users, cluster configuration is rarely manipulated at run time - if any changes are made to it, it would probably be a permanent change, meaning a code change would be required.

I can get rid of its config for now then, and wait until a user explicitly asks for this run-time configuration of their jobs.

schrockn commented on python_modules/libraries/dagster-databricks/dagster_databricks/ops.py:
```diff
@@ -135,3 +138,224 @@ def create_ui_url(databricks_client, op_config):
+    @op(
+        ins=(ins or {"start_after": In(Nothing)}),
+        out=out,
+        config_schema={
+            "poll_interval_sec": Field(
+                float,
+                description="Check whether the Databricks Job is done at this interval.",
+                default_value=_DEFAULT_POLL_INTERVAL,
+            ),
+            "max_wait_time_sec": Field(
+                float,
+                description=(
+                    "If the Databricks Job is not complete after this length of time, raise an"
+                    " error."
+                ),
+                default_value=_DEFAULT_RUN_MAX_WAIT_TIME_SEC,
+            ),
+        },
+        required_resource_keys={"databricks"},
+        tags={"kind": "databricks"},
```
in terms of parameterizing config, what I think *would* make sense is to make these default values parameterizable, which is right way to get "best of both worlds" here.

schrockn replied:
And I think the approach is _preferable_ to calling `configured`

rexledesma replied:
Addressed in https://github.com/dagster-io/dagster/pull/12492/commits/9c42bfdc217cc32e0755af7eca2accb921765470

schrockn commented:
Last question in this one is about parameterizing `ins` and `outs` here. This is a case where the asset variant of this makes more sense (upstream asset keys and the asset_keys materialized by the databricks job) but the op mental model is less great.

rexledesma commented:
> Last question in this one is about parameterizing `ins` and `outs` here. This is a case where the asset variant of this makes more sense (upstream asset keys and the asset_keys materialized by the databricks job) but the op mental model is less great.

Alright, we'll also divest from this functionality and only expose it in the `multi_asset` factory version of these APIs. There was some thinking here that if we wanted the user to migrate from the op factory to its asset version, they would use `AssetDefinition.from_op`. But instead, we'll assume that they just migrate wholesale to the `multi_asset` factory version.
